power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/iterative_linear_se_solver.hpp

```diff
@@ -355,9 +355,9 @@ template <bool sym> class IterativeLinearSESolver {
 template class IterativeLinearSESolver<true>;
 template class IterativeLinearSESolver<false>;
 
-} // namespace math_model_impl
+} // namespace math_model_impl::ilse_solver_impl
 
-template <bool sym> using IterativeLinearSESolver = math_model_impl::IterativeLinearSESolver<sym>;
+template <bool sym> using IterativeLinearSESolver = math_model_impl::ilse_solver_impl::IterativeLinearSESolver<sym>;
```

Comment:
if you want to, you can create a namespace alias using `namespace ilse = ilse_solver_impl`

in addition, I do not like the duplication in `math_model_impl` and then ALSO `ilse_solver_impl`.

I would just replace the entire `math_model_impl` namespace with `math_model::impl` and then the `ilse_solver_impl` just becomes `math_model::impl::ilse`. Anything that resides in something called `impl` by definition is also `impl`, so no repetition is necessary

Idem for nrse of course

Reply:
So 3 namespaces? What things do math_model cover? Or directly nest?

Reply:
you can see a namespace as a sort of json-like list that can contain function names and other namespaces. having multiple nested ones (as long as the list doesn't become TOO large) is perfectly OK.

i see that `math_model` actually wouldn't contain anything (i thought the `math_solver` was part of that namespace as well but it isn't)

i personally would have a namespace per directory in which the headers reside (e.g.: `power_grid_model`, `power_grid_model::main_core`, `power_grid_model::component`, `power_grid_model::math_solver` and `power_grid_model::auxiliary`), and anything that does not need to be exposed outside of certain headers or the directory may reside in `<directory>::detail` (helper functions) or `<directory>::impl` (implementation details of something that IS exposed)

In that line, it would theoretically be fine to have just `power_grid_model::math_solver` but there are so many implementation details per solver that it starts to make sense to put them in separate namespaces per header as well. I have a preference to not have too many class member functions (even if they are `static`). It promotes reasonings like "oh we have it available anyways, so I don't need to make this function `static`" which in turn results in spaghetti-code. Instead, I prefer to mimic the same behaviour as `ClassName::member_function` by `ns::static_function`. For both the compiler and the user, the behaviour is exactly the same (aside from `public`/`protected`/`private`, of course).

If you rename `math_model_impl` to just `math_solver`, the rest should be OK and you indeed don't need `math_model::impl`